### DISCUSSION TITLE
Get rid of the option "enabled"

### DISCUSSION THREAD

**Discussion by artem-malko** [TanStack/query#4765] [Ideas] [answered]
The main idea here is useQuery is a hook, one of the hook's rule is "All the hooks should be executed unconditionally". But useQuery is quite special here) 

The second idea is — difficulties of a query state management during render. I mean, that we have to handle not only error/loading/success, but paused too. 

Cause of that flag we have to develop some ways, how to use it with TypeScript https://tkdodo.eu/blog/react-query-and-type-script#type-safety-with-the-enabled-option. But as for me (sorry  @TkDodo) it's just a spike, not a solution. 

And sometimes "enabled" is like a fast patch in all of the cases in my opinion. Do not think, just use enabled. I know, that it is a problem of developers in the first place, but, as I think, TanStack query should learn how to write good code =)

### There is no params for a query

Actually you have such example in the docs: https://tanstack.com/query/v4/docs/react/guides/disabling-queries#lazy-queries
But, of the top of my head, this problem could be solved by different components:
- In the main component we have `[filter, setFilter]`
- In the child component we have useQuery.

So, we need to render the child component with useQuery only if we have any truthy value in `filter`. By the way, this approach allows to write more reliable applications.

### Conditionals queries

```ts
const q1 = useQuery(['q1']);
const q2 = useQuery(['q2', q1.data]);
```

Actually this problem has the same solutions — conditional rendering =)

I'm glad to see another examples, when we need such option. And I wanna discuss it here.

So, right now I have 2 questions and 1 idea. Questions:

1. Why do we need to have that option?
2. Is it possible to get rid of it?

And the idea — to get rid of it =)

**Comment by artem-malko** [comment]
Hey, @TkDodo @Ephem what do you think?)

**Comment by TkDodo** [comment, marked as answer]
> Why do we need to have that option?

Because it's not possible to call hooks conditionally. It's also not always easily possible or desired to split things up into multiple components, especially when writing custom hooks. Consider:

```
const useSomething = () => {
  const userId = useSelector(state => state.user?.id)
  return useQuery({
    queryKey: ['something', userId],
    queryFn: () => fetchSomething(userId),
    enabled: !!userId
  })
}
```

sure you _can_ split that up into multiple components, but that would mean prop-drilling something that you read from a global client state manager (like redux in this example) and you can't really use custom hooks composition nicely.

it also occurs frequently when reading `params` with `useRouter` from `nextJs`, because those params are undefined on the first render. They return a `router.isReady` property that is frequently used to disable a query.

**Reply by artem-malko** [reply]
A custom hook is a real problem, you're right. But, as you said, "it's not possible to call hooks conditionally", but you do. And for the biggest part of developers its much easy to just use that "enable" option, instead of other solutions =( 

So, may be it's a good idea to be honest in API? I mean, if you need to call something conditionally (imperatively), so there could be something like, `fetchIfNeeded` function. I know, we have queryClient.fetchQuery function. But, this function will be called in all conditions. So, a new one, like `fetchIfNeeded` can be called only if it is needed.

Maybe I'm too strict with "enabled", but as I think, this option breaks the main rule of hooks. And `useSWR` doesn't have such options, as I know.

**Reply by TkDodo** [reply]
> But, as you said, "it's not possible to call hooks conditionally", but you do.

We don't - that wouldn't be allowed by the rules of hooks. The hook and all underlying hooks like `useSyncExternalStore` are always called. What enabled does is it just makes sure that this one observer doesn't fetch data on its own. The observer will still receive updates that are put into the cache by other means, like calling `setQueryData`, `prefetchQuery`, or having another observer that observes the same `queryKey` and that _isn't_ disabled.

If we could call hooks conditionally, we could have an api that would give us the chance to call two different hooks depending on the condition:

```
const enabled = !!userId
const query = { queryKey, queryFn }

const result = enabled ? useQuery(query) : useQueryData(query)
```

where `useQueryData` would give you the data for the query that is currently in the cache and subscribe you to updates that will happen for that key, because technically, that is what the `enabled` option does.

> I know, we have queryClient.fetchQuery function. But, this function will be called in all conditions.

Not sure what this means exactly, but you can pass a `staleTime` to `fetchQuery`, in which case it will only fetch if there is no data available in the cache that is older than this time.

**Reply by TkDodo** [reply]
In swr, you can pass in `null` as the `key` to achieve a similar thing as `enabled`

**Reply by artem-malko** [reply]
Yeah, you're right, we can pass  a `staleTime` to `fetchQuery`. So, `fetchQuery` can be a solution in a custom hook. Ok, I understand your main idea, that `enabled` is just to disable observer for the current query.

By the way, in SWR I can pass a falsy value as a queryFn to achieve `enabled: false`. And it's much more conveniently, cause you don't need to return rejected Promise and you don't have any problems with TypeScript.

```ts
type Param = number | undefined;

const customQuery = (param: Param) => useQuery({
  key: ['someKey', param],
  queryFn: param ?? () => { fetch(`/something/${param}`) },
});
```

So, if param is undefined, useQuery won't make any requests.

**Reply by TkDodo** [reply]
except that you can also use `refetch` to bypass `enabled` :)

**Reply by artem-malko** [reply]
> except that you can also use refetch to bypass enabled :)

I didn't get it. Could you explain, pls?)

**Reply by artem-malko** [reply]
`refetch` has to work the same way. No `Params` — no request.

**Reply by TkDodo** [reply]
no, if you have the example from above:

```
type Param = number | undefined;

const customQuery = (param: Param) => useQuery({
  key: ['someKey', param],
  queryFn: param ?? () => { fetch(`/something/${param}`) },
});
```

and you call `customQuery.refetch()`, it _will_ refetch. However, `queryClient.invalidateQueries` or `queryClient.refetchQueries` will _not_ target this query.

The reason is that you can opt-out into manual mode by doing `enabled: false` and then just fetch with `refetch()` if you want to, even though it's not recommended.

This is [documented here](https://tanstack.com/query/v4/docs/react/guides/disabling-queries).

> refetch returned from useQuery can be used to manually trigger the query to fetch.

**Reply by artem-malko** [reply]
Oh, ok. I understand. It goes to deep right now with `enable`.
